File: listing-dashboard/remove_watermark.py
#!/usr/bin/env python3
"""
Remove the bottom-right sparkle watermark (✦) from listing photos.
Usage:
  python3 remove_watermark.py photo.jpg
  python3 remove_watermark.py *.jpg
  python3 remove_watermark.py /path/to/folder/
"""
import sys, os, glob
from PIL import Image
import numpy as np


def remove_watermark(path, wm_size=65):
    """
    Patches the bottom-right corner (wm_size x wm_size px) by replacing
    bright/white sparkle pixels with a sampled patch from just above it.
    Works on any image with a light-colored corner watermark.
    """
    img = Image.open(path).convert('RGBA')
    arr = np.array(img, dtype=np.float32)
    h, w = arr.shape[:2]

    # The patch we'll use as replacement (same size, directly above the watermark)
    ref = arr[h - wm_size*2 : h - wm_size, w - wm_size : w].copy()

    # The watermark corner
    corner = arr[h - wm_size : h, w - wm_size : w].copy()

    # Detect sparkle: bright pixels significantly lighter than surroundings
    # The sparkle is white/near-white against any background
    brightness = corner[:, :, :3].mean(axis=2)
    ref_brightness = ref[:, :, :3].mean(axis=2)
    avg_bg = ref_brightness.mean()

    # Mask: pixels brighter than background by more than 40 and brighter than 180
    mask = (brightness > min(avg_bg + 40, 200)) & (brightness > 170) & (corner[:, :, 3] > 50)

    # Also grab a 10px soft border around the mask for smoother blending
    from PIL import ImageFilter
    mask_img = Image.fromarray((mask * 255).astype(np.uint8))
    soft_mask = np.array(mask_img.filter(ImageFilter.GaussianBlur(radius=3))) / 255.0

    # Blend: replace masked pixels with reference
    for c in range(4):
        corner[:, :, c] = corner[:, :, c] * (1 - soft_mask) + ref[:, :, c] * soft_mask

    arr[h - wm_size : h, w - wm_size : w] = corner

    result = Image.fromarray(arr.astype(np.uint8), 'RGBA').convert('RGB')

    # Save — preserve original format
    ext = os.path.splitext(path)[1].lower()
    save_kwargs = {'quality': 95, 'subsampling': 0} if ext in ('.jpg', '.jpeg') else {}
    result.save(path, **save_kwargs)
    print(f'  ✓  {os.path.basename(path)}')


def main():
    if len(sys.argv) < 2:
        print(__doc__)
        sys.exit(1)

    targets = []
    for arg in sys.argv[1:]:
        if os.path.isdir(arg):
            for ext in ('*.jpg', '*.jpeg', '*.png', '*.webp'):
                targets.extend(glob.glob(os.path.join(arg, ext)))
                targets.extend(glob.glob(os.path.join(arg, ext.upper())))
        else:
            targets.extend(glob.glob(arg))

    if not targets:
        print('No image files found.')
        sys.exit(1)

    print(f'Removing watermarks from {len(targets)} photo(s)...')
    for p in sorted(set(targets)):
        try:
            remove_watermark(p)
        except Exception as e:
            print(f'  ✗  {os.path.basename(p)}: {e}')

    print('Done.')


if __name__ == '__main__':
    main()

File: listing-dashboard/server/removeWatermark.js
/**
 * Watermark removal using sharp (fast Node.js image processor).
 * Covers the Zillow/Trulia 4-pointed sparkle ✦ in the bottom-right corner
 * by sampling a patch of pixels from just above/left of the watermark area
 * and compositing it over the mark.
 */

let sharp;
try { sharp = require('sharp'); } catch (_) { sharp = null; }

const { execSync } = require('child_process');
const path = require('path');
const fs = require('fs');

/**
 * Remove the bottom-right sparkle watermark from an image file (in-place).
 * Uses sharp if available, otherwise falls back to Python/Pillow.
 */
async function removeWatermark(filepath) {
  if (sharp) {
    return removeWithSharp(filepath);
  }
  return removeWithPython(filepath);
}

async function removeWithSharp(filepath) {
  const image = sharp(filepath);
  const meta = await image.metadata();
  const { width, height } = meta;

  // Watermark region: bottom-right corner, roughly 54x54px
  const WM = 54;
  const left = width - WM;
  const top  = height - WM;

  // Extract a patch from just above the watermark (same width, same height)
  const patch = await sharp(filepath)
    .extract({ left, top: top - WM, width: WM, height: WM })
    .toBuffer();

  // Composite the patch over the watermark region
  await sharp(filepath)
    .composite([{ input: patch, left, top }])
    .toFile(filepath + '.tmp');

  fs.renameSync(filepath + '.tmp', filepath);
  return { ok: true, method: 'sharp' };
}

function removeWithPython(filepath) {
  const script = `
import sys
from PIL import Image
import numpy as np

path = sys.argv[1]
img = Image.open(path).convert('RGBA')
arr = np.array(img, dtype=np.float32)
h, w = arr.shape[:2]

WM = 60  # watermark region size in px

# Sample a reference patch from just above the watermark
ref = arr[h - WM*2 : h - WM, w - WM : w].copy()

# Detect the sparkle: white pixels (R,G,B all > 200) in the corner
corner = arr[h - WM : h, w - WM : w]
mask = (corner[:,:,0] > 180) & (corner[:,:,1] > 180) & (corner[:,:,2] > 180) & (corner[:,:,3] > 100)

# Replace masked pixels with the reference patch pixels
for c in range(4):
    corner[:,:,c][mask] = ref[:,:,c][mask]

arr[h - WM : h, w - WM : w] = corner

result = Image.fromarray(arr.astype(np.uint8), 'RGBA').convert('RGB')
result.save(path, quality=95)
print('ok')
`.trim();

  const scriptPath = filepath + '.wm.py';
  fs.writeFileSync(scriptPath, script);
  try {
    execSync(`python3 "${scriptPath}" "${filepath}"`, { timeout: 30000 });
    return { ok: true, method: 'python' };
  } finally {
    try { fs.unlinkSync(scriptPath); } catch (_) {}
  }
}

module.exports = { removeWatermark };
